### ISSUE TITLE
Nominal Interest Rate is not updating for all saving accounts after Changes has been made in Saving Product’s “ Nominal Annual Interest Rate”

### ISSUE THREAD

**Issue by abiyats** [openMF/community-app#3017]
## Description
This is the scenario I need to report. 

Lets say I have created a Saving Account #000001 with Saving Product “Mandatory Saving”. Initially, The Nominal Annual Interest Rate for Mandatory Saving was 5%.

After change of management or other reasons by the operation, one may need to increase the Interest Rate to 7%.

The error is :

 The saving accounts created before the amendment of new interest rate are not updating their Nominal Interest Rate. It is 5% when checked in viewsavingaccout page.

But those saving accounts created after the amendment are created with 7%.

N.B. I have checked the math with calculator.

## Steps to Reproduce

Step1: Create a saving Product (Mandatory Saving) with Nominal Annual Interest Rate of 5% and fill all the rest as required.

Step2: Create a a client with a saving account #000001 with (Mandatory Saving ) product.

Step3: Check saving account details for the client created in step 2. (Make sure the saving account's Nominal Interest Rate=5)

Step3: Run Scheduler Job. (take the system date a day or week forward)

Step4: Go to PRODUCTS (Mandatory Saving Product) and change the Nominal Annual Interest Rate to 7%.

Step5: Repeat step 2.

Step 6: Check saving account details for the client created in step 5 #000002 (Make sure the saving account's Nominal Interest Rate=7

Step 7: Both clients #000001 & #000002 in step 2 and step 5 must now have the same interest rate 7%. (meaning that there should no be any saving account with 5% under that particular saving product - Mandatory Saving)

## Expected Behaviour
I believe at least from the amendment date forwards **ALL** saving accounts under that specific Saving Product (no matter when they were created) should update their Nominal Annual Interest Rate.

## Actual Behaviour
**Only those** saving accounts created after the amendment of the new interest rate #000002 have the new interest rate (7%). But those saving accounts created before the change #000001 still have 5% interest rate which should be changed from 5% - 7% after the amendment.

## Settings
- Mifos X version: 18.03.xx
- Browser used: Google Chrome, Mozilla Firefox

## Screenshots, if any

![5](https://user-images.githubusercontent.com/39126615/47849961-982a6f80-dde4-11e8-9794-9b6b32a8b00f.PNG)
![sp](https://user-images.githubusercontent.com/39126615/47849964-995b9c80-dde4-11e8-90f5-7bb2af96c562.PNG)

![7](https://user-images.githubusercontent.com/39126615/47849963-98c30600-dde4-11e8-9ca0-fdb280e53faf.PNG)

**Comment by santoshconflux** [comment]
@abiyats , any changes made in product level (loan,savings, etc) will affect those accounts which are created after the change. It should not affect the existing accounts. It's correct behaviour.

**Comment by abiyats** [comment]
Hello Santosh,

Okay. I understand brother, But what if the bank wants to increase interest
rates for all accounts? That was my question.

BTW, I have done it in database level. Selected all loans with activation
date less than the interest rate increment date, updated the nominal annual
interest field with the new rate.

And can we add a check box under the loan product page, to alow the change
for existing loans, by which it triggers the DB to select and update all
loan accounts with activation date are less than the ammendment sate?

Regards,
Abiy A.

**Comment by abiyats** [comment]
Dear Santosh,

I was updating the contracts using the following sql command.

For Saving:
Assume, previous interest rate was 5%, and wanted to change to 7%.

UPDATE `mifostenant-default`.`m_savings_account` SET
`nominal_annual_interest_rate`='7.000000' WHERE  `activatedon_date` <
20181214;

For Loans:
Assume, previous interest rate was 10%, and wanted to change to 15%.

UPDATE `mifostenant-default`.`m_loan` SET
`annual_nominal_interest_rate`='15.000000' WHERE
`activatedon_date`<20181214;
UPDATE `mifostenant-default`.`m_loan` SET
`nominal_interest_rate_per_period`='15.000000' WHERE
`activatedon_date`<20181214;

Thus, meaning this will select all saving accounts created before 20181214.
B/c all other accounts can be affected after the update. (It will get the
date from the system date)
So, my recommendation is can we add a check box type to allow all
saving/loan accounts to update the interest rates with the new rate as per
the clients requirement? Just like the attachment... And the check box will
trigger the SQL command.


Thanks & Regards,
Abiy A.

**Comment by santoshconflux** [comment]
@abiyats , Consider the one real scenario , you have already posted the interest 13% to the client's savings account after 1st quarter. Now you're updating product to the new interest at 15% and you made effective from savings activation date. Now all the transactions related to it must be updated and every journal entries as well. I think lot of tables get affected. I don't think it's easy  job. Let me confirm this with one of our developer.

**Comment by abiyats** [comment]
@Santosh,

Yeah you are right, we need to make it take an effect from the Rate
increment Date, not the saving activation date, because in that case it
will have to do backdate transactions which will affect journal entries.
(This must be done by local development by the client, and the system
should not be accountable for such actions) I agree with you.

But if these changes take effect right after the change date, it will start
calculating the interest rate for the savings/lon accounts with the new
rate after the ammendment date.

Assume if we are incrementing the rate from 13% to 15% on 14 December 2018,
 the changes must take effect for all accounts that are created after it
and all the Nominal Interest Rate of all existing contracts will be
updated, then the system will calculate on the next COB (Close of Business)
- Cron Job, at night of 14 December 2018.

Regards,
Abiy A.
